build/yaml/steps/codeanalysis/csharp.yaml

```diff
@@ -10,6 +10,6 @@ steps:
     continueOnError: true
     rulesetName: Custom
     msBuildArchitecture: x64
-    customRuleset: DevDivRoslynRequired.ruleset
+    customRuleset: $(Build.SourcesDirectory)\build\CodeAnalysis\Sdl.Recommended.Error.ruleset
```

Comment:
> $(Build.SourcesDirectory)\build\CodeAnalysis\Sdl.Recommended.Error.ruleset [](http://example.com/codeflow?start=18&length=75)

Are we sure that these two rulesets are the same?

Reply:
It is not the same; the prior behavior was referring to a non-existent rule set. There is no DevDivRoslynRequired.ruleset in the repo.

Compared to the hypothetical DevDivRoslynRequired, the new ruleset should be more restrictive (more enabled rules, and all rules at the same or more restrictive level).

Reply:
I'm not certain that the ruleset has to be in the repository. It might be included with the guardian toolset that gets downloaded for the task. I would have to look at the documentation for the task again, though.

Reply:
The documentation says that the customRuleset should be "An accessible path to a ruleset to use." Using a built-in ruleset is only supported by specifying the resultName input. However, since we want the build analyzers and this task to agree on the ruleset, we should be using the path to the custom ruleset.

Reply:
I dug into this more this morning to figure out what was happening.

Prior to this change, there were a few code analysis warnings that the RoslynAnalyzers task would flag, for example, CA1062. If you check the actual DevDivRoslynRequired.ruleset or the Sdl.Recommended.Error.ruleset, this rule is set to None in both cases. Also, if you check the RoslynAnalyzers build, there are a bunch of "Could not find rule set file C:\a\1\s\DevDivRoslynRequired.ruleset" warnings. This leads me to believe that the RoslynAnalyzers task was just running all possible rules.

I wrote a quick tool to compare the Sdl.Recommended.Error.ruleset and DevDivRoslynRequired.ruleset files and found that the SDL one is a superset of the DevDiv one, with the exception of a handful of internal rules (they start with IA). Because they are internal, they require a special NuGet package to enable those rules, which I don't think we want to put up on the dependency feed. So I think we need to use the SDL ruleset.

I don't see any precedence for adding the DevDivRoslynRequired.rulset to any of the GitHub repos in the microsoft or dotnet orgs.